drivers/at/at.c

```diff
@@ -0,0 +1,238 @@
+/*
+ * Copyright (C) 2017 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+#include <errno.h>
+#include <string.h>
+
+#include "at.h"
+#include "fmt.h"
+#include "isrpipe.h"
+#include "periph/uart.h"
+#include "xtimer.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+#ifndef AT_PRINT_INCOMING
+#define AT_PRINT_INCOMING (0)
+#endif
+
+int at_dev_init(at_dev_t *dev, uart_t uart, uint32_t baudrate, char *buf, size_t bufsize)
+{
+    dev->uart = uart;
+    isrpipe_init(&dev->isrpipe, buf, bufsize);
+    uart_init(uart, baudrate, (uart_rx_cb_t) isrpipe_write_one,
+              &dev->isrpipe);
+
+    return 0;
+}
+
+int at_expect_bytes(at_dev_t *dev, const char *bytes, size_t len, uint32_t timeout)
+{
+    while (len) {
+        char c;
+        int res;
+        if ((res = isrpipe_read_timeout(&dev->isrpipe, &c, 1, timeout)) == 1) {
+            if (AT_PRINT_INCOMING) {
+                print(&c, 1);
+            }
+            if (!(c == *bytes++)) {
+                return -1;
+            }
+            len--;
+        }
+        else {
+            return res;
+        }
+    }
+
+    return 0;
+}
+
+void at_send_bytes(at_dev_t *dev, const char *bytes, size_t len)
+{
+    uart_write(dev->uart, (const uint8_t *)bytes, len);
+}
+
+int at_send_cmd(at_dev_t *dev, const char *command, uint32_t timeout)
+{
+    unsigned cmdlen = strlen(command);
+
+    uart_write(dev->uart, (const uint8_t *)command, cmdlen);
+    uart_write(dev->uart, (const uint8_t *)AT_END_OF_LINE, sizeof(AT_END_OF_LINE) - 1);
+
+    if (at_expect_bytes(dev, command, cmdlen, timeout)) {
+        return -1;
+    }
+
+    if (at_expect_bytes(dev, AT_END_OF_LINE "\r\n", sizeof(AT_END_OF_LINE) + 1, timeout)) {
```

Comment:
maybe I missunderstand the line above, but `AT_END_OF_LINE` is
```c
#define AT_END_OF_LINE "\r"
```
So the line above does make a string concatenation, leading to `\r\r\n`, right?!